File: Very Hard/High-rise-buildings/High-rise-buildings.java
class Solution {

    public static void main(String args[]) {
        var in = new java.util.Scanner(System.in);
        int n = in.nextInt();

        // Some data structures
        int[][] grid = new int[n][n];
        int[][] clues = new int[4][n];
        boolean[][] mustStay = new boolean[n][n];

        // Read clues
        for (int i = 0; i < 4; i++)
            for (int j = 0; j < n; j++)
                clues[i][j] = in.nextInt();

        // Read starting cells
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++) {
                grid[i][j] = in.nextInt();
                mustStay[i][j] = (grid[i][j] != 0);
            }

        // Solve and print
        solve(grid, mustStay, clues, n);
        print(grid, n);
    }

    // Very simple backtracking, same as a sudoku solver but different validation function
    private static boolean solve(int[][] grid, boolean[][] mustStay, int[][] clues, int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] != 0) {
                    continue;
                } else {
                    for (int k = 1; k <= n; k++) {
                        if (valid(k, grid, clues, i, j, n)) {
                            grid[i][j] = k;
                            if (solve(grid, mustStay, clues, n)) {
                                return true;
                            } else if (!mustStay[i][j]) {
                                grid[i][j] = 0;
                            }
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }

    // Check if a number k is possible at grid[i][j]
    private static boolean valid(int k, int[][] grid, int[][] clues, int i, int j, int n) {
        // Check row and column
        for (int z = 0; z < n; z++) {
            if (grid[z][j] == k)
                return false;
            if (grid[i][z] == k)
                return false;
        }
        // Check visibility clues
        // NORTH
        int N = clues[0][j];
        int[] arrayN = new int[n];
        boolean[] arrayMustStayN = new boolean[n];
        for (int z = 0; z < n; z++) {
            arrayN[z] = (z == i) ? k : grid[z][j];
            arrayMustStayN[z] = arrayN[z] != 0;
        }
        if (!validVis(arrayN, arrayMustStayN, N, n))
            return false;
        // SOUTH
        int S = clues[3][j];
        int[] arrayS = new int[n];
        boolean[] arrayMustStayS = new boolean[n];
        for (int z = 0; z < n; z++) {
            arrayS[z] = (z == n - 1 - i) ? k : grid[n - 1 - z][j];
            arrayMustStayS[z] = arrayS[z] != 0;
        }
        if (!validVis(arrayS, arrayMustStayS, S, n))
            return false;
        // WEST
        int W = clues[1][i];
        int[] arrayW = new int[n];
        boolean[] arrayMustStayW = new boolean[n];
        for (int z = 0; z < n; z++) {
            arrayW[z] = (z == j) ? k : grid[i][z];
            arrayMustStayW[z] = arrayW[z] != 0;
        }
        if (!validVis(arrayW, arrayMustStayW, W, n))
            return false;
        // EAST
        int E = clues[2][i];
        int[] arrayE = new int[n];
        boolean[] arrayMustStayE = new boolean[n];
        for (int z = 0; z < n; z++) {
            arrayE[z] = (n - 1 - z == j) ? k : grid[i][n - 1 - z];
            arrayMustStayE[z] = arrayE[z] != 0;
        }
        if (!validVis(arrayE, arrayMustStayE, E, n))
            return false;
        return true;
    }

    // Is there at least 1 valid way to fill the array with the given visibility clue?
    private static boolean validVis(int[] array, boolean[] arrayMustStay, int vis, int n) {
        for (int i = 0; i < n; i++) {
            if (array[i] != 0)
                continue;
            for (int j = 1; j <= n; j++)
                if (notInArray(array, j, n)) {
                    array[i] = j;
                    if (validVis(array, arrayMustStay, vis, n))
                        return true;
                    else if (!arrayMustStay[i])
                        array[i] = 0;
                }
            return false;
        }
        return (vis == countVis(array));
    }

    private static boolean notInArray(int[] array, int num, int n) {
        for (int i = 0; i < n; i++)
            if (array[i] == num)
                return false;
        return true;
    }

    private static int countVis(int[] array) {
        int count = 0, max = -1;
        for (int i = 0; i < array.length; i++)
            if (array[i] > max) {
                count++;
                max = array[i];
            }
        return count;
    }

    private static void print(int[][] grid, int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++)
                System.out.print(grid[i][j] + (j != n - 1 ? " " : ""));
            System.out.println();
        }
    }
}

File: Very Hard/High-rise-buildings/High-rise-buildings.c
#include <stdio.h>
#include <stdlib.h>

int solve();
int valid(int k, int i, int j);
int validVis(int *array, int *mustStayD, int vis);
int notInArray(int *array, int num);
int countVis(int *array);
void printGrid();

int n;
int **grid;
int **clues;
int **mustStay;

int main() {
  scanf("%d", &n);

  // Read clues
  clues = (int **)malloc(sizeof(int *) * 4);
  for (int i = 0; i < 4; i++) {
    clues[i] = (int *)malloc(sizeof(int) * n);
    for (int j = 0; j < n; j++) {
      scanf("%d", clues[i] + j);
    }
  }

  // Read initial grid
  grid = (int **)malloc(sizeof(int *) * n);
  mustStay = (int **)malloc(sizeof(int *) * n);
  for (int i = 0; i < n; i++) {
    grid[i] = (int *)malloc(sizeof(int) * n);
    mustStay[i] = (int *)malloc(sizeof(int) * n);
    for (int j = 0; j < n; j++) {
      scanf("%d", grid[i] + j);
      mustStay[i][j] = (grid[i][j] != 0);
    }
  }

  solve();
  printGrid();
}

int solve() {
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      if (grid[i][j] != 0)
        continue;
      for (int k = 1; k <= n; k++) {
        if (valid(k, i, j)) {
          grid[i][j] = k;
          if (solve()) {
            return 1;
          } else if (!mustStay[i][j]) {
            grid[i][j] = 0;
          }
        }
      }
      return 0;
    }
  }
  return 1;
}

int valid(int k, int i, int j) {
  for (int z = 0; z < n; z++) {
    if (grid[z][j] == k)
      return 0;
    if (grid[i][z] == k)
      return 0;
  }
  int N = clues[0][j];
  int arrayN[n];
  int mustStayN[n];
  for (int z = 0; z < n; z++) {
    arrayN[z] = (z == i) ? k : grid[z][j];
    mustStayN[z] = arrayN[z] != 0;
  }
  if (!validVis(arrayN, mustStayN, N))
    return 0;

  int S = clues[3][j];
  int arrayS[n];
  int mustStayS[n];
  for (int z = 0; z < n; z++) {
    arrayS[z] = (z == n - 1 - i) ? k : grid[n - 1 - z][j];
    mustStayS[z] = arrayS[z] != 0;
  }
  if (!validVis(arrayS, mustStayS, S))
    return 0;

  int W = clues[1][i];
  int arrayW[n];
  int mustStayW[n];
  for (int z = 0; z < n; z++) {
    arrayW[z] = (z == j) ? k : grid[i][z];
    mustStayW[z] = arrayW[z] != 0;
  }
  if (!validVis(arrayW, mustStayW, W))
    return 0;

  int E = clues[2][i];
  int arrayE[n];
  int mustStayE[n];
  for (int z = 0; z < n; z++) {
    arrayE[z] = (z == n - 1 - j) ? k : grid[i][n - 1 - z];
    mustStayE[z] = arrayE[z] != 0;
  }
  if (!validVis(arrayE, mustStayE, E))
    return 0;
  return 1;
}

int validVis(int *array, int *mustStayD, int vis) {
  for (int i = 0; i < n; i++) {
    if (array[i] != 0)
      continue;
    for (int j = 1; j <= n; j++)
      if (notInArray(array, j)) {
        array[i] = j;
        if (validVis(array, mustStayD, vis))
          return 1;
        else if (!mustStayD[i])
          array[i] = 0;
      }
    return 0;
  }
  return vis == countVis(array);
}

int notInArray(int *array, int num) {
  for (int i = 0; i < n; i++)
    if (array[i] == num)
      return 0;
  return 1;
}

int countVis(int *array) {
  int count = 0, max = -1;
  for (int i = 0; i < n; i++)
    if (array[i] > max) {
      count++;
      max = array[i];
    }
  return count;
}

void printGrid() {
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      printf("%d", grid[i][j]);
      if (j != n - 1)
        printf(" ");
    }
    printf("\n");
  }
}
